### PR TITLE
API changes for push in libgit2 0.22

### PR DESCRIPTION
**Pull request by phatblat**
Resurrect and update the push functionality.

Note that I could not get either of the libgit2 builds to work without defining LIBSSH2_INCLUDE_DIR in the [build scripts](https://github.com/phatblat/objective-git/compare/ben/pr/libgit2-build).

### CODE REVIEW

File: ObjectiveGit/GTRepository+RemoteOperations.h [hunk 1]
```diff
@@ -45,4 +45,34 @@ extern NSString *const GTRepositoryRemoteOptionsCredentialProvider;
 /// Retruns an array with GTFetchHeadEntry objects
 - (NSArray *)fetchHeadEntriesWithError:(NSError **)error;
 
+#pragma mark - Push
+
+/// Push a single branch to a remote.
+///
+/// branch        - The branch to push.
```
Comment (joshaber) [inline ObjectiveGit/GTRepository+RemoteOperations.h]: Could you document that this and other properties cannot be nil?

File: ObjectiveGit/GTRepository+RemoteOperations.m [hunk 2]
```diff
@@ -159,4 +160,99 @@ - (NSArray *)fetchHeadEntriesWithError:(NSError **)error {
...
+	// Build refspecs for the passed in branches
+	refspecs = [NSMutableArray arrayWithCapacity:branches.count];
+	for (GTBranch *branch in branches) {
+		// Default remote reference for when branch doesn't exist on remote - create with same short name
+		NSString *remoteBranchReference = [NSString stringWithFormat:@"refs/heads/%@", branch.shortName];
+
+		BOOL success = NO;
+		GTBranch *trackingBranch = [branch trackingBranchWithError:error success:&success];
+
+		if (success && trackingBranch) {
```
Comment (joshaber) [inline ObjectiveGit/GTRepository+RemoteOperations.m]: Style: explicit comparison to `nil` for `trackingBranch`.

File: ObjectiveGitTests/GTRemotePushSpec.m [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+			[NSFileManager.defaultManager removeItemAtURL:remoteRepoURL error:&error];
+			expect(error).to(beNil());
+			[NSFileManager.defaultManager removeItemAtURL:localRepoURL error:&error];
+			expect(error).to(beNil());
+			error = NULL;
+		});
+
+		context(@"when -pushBranch: is given invalid parameters", ^{
+			it(@"needs a non-nil branch", ^{
+				XCTAssertThrowsSpecificNamed([localRepo pushBranch:nil toRemote:remote withOptions:nil error:&error progress:NULL], NSException, NSInternalInconsistencyException, @"should throw NSInternalInconsistencyException");
```
Comment (joshaber) [inline ObjectiveGitTests/GTRemotePushSpec.m]: We don't typically test assertions.

**Comment by joshaber** [comment]
Awesome, thanks so much for re-doing this! :metal:

**Comment by phatblat** [comment]
Sure thing. Ready for rerereview.

**Pull request merged**